Improve webpack documentation

tgandrews opened the pull request:
This aims to resolve #5679

This improves the documentation for production mode, but does not mention about development mode.

cody commented:
Getting docs for webpack is very nice. Thank you.

Using separate confics for production and development seems very common, but I don't understand why. There are only few differences and the we already have to use the `-p` flag anyway. I use something like this:
```js
var isProduction = process.argv.indexOf('-p') >= 0;

if (isProduction) {
  plugins = [
    new webpack.optimize.UglifyJsPlugin({compress: {warnings:false}}),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': '"production"'
    })
  ];
} else {
  plugins = [
    new webpack.DefinePlugin({
       'process.env.NODE_ENV': '"development"'
    })
  ];
}
```

tgandrews commented:
The differences in my current dev and prod mode include:
- Image crushing in prod
- File hashing in the output filename in prod
- Minification in prod
- Hot module reloading in dev
- Source maps in dev

That seems enough to warrant a separate config file instead of having a conditionals all over the place.

cody commented:
@tgandrews 
The docs should at least mention the dev build. Also I would aim for an example that works out of the box, but that is as simple as possible.

I think hashing of the file name is not needed here.

I am not a dev, therefore cc @zpao for better feetback. :smile:

tgandrews commented:
@cody 
I have simplified the prod config (removed the hahsing) and added changes that could be made for the dev build. I agree simpler is better. The original issue seemed to be regarding the complexity of getting the production config set up so I have made that the main focus.

brigand commented on docs/docs/getting-started.md:
```diff
@@ -41,9 +41,63 @@ $ npm install --save react react-dom babel-preset-react
-> Note:
->
-> If you are using ES2015, you will want to also use the `babel-preset-es2015` package.
+Ensure you have a `.babelrc` including the react preset.
```
Setting up a .babelrc isn't trivial to people that haven't used it before, especially in babel 6. A basic config should be included here.

andreypopp replied:
Can the babel config be specified in `webpack.config.js` (in loader config)? I think having that is simpler. People could learn about `.babelrc` later on.

tgandrews replied:
I want to leave things as jumping off points, otherwise we should be providing a full skeleton.

brigand commented on docs/docs/getting-started.md:
````diff
@@ -41,9 +41,63 @@ $ npm install --save react react-dom babel-preset-react
+};
+```
+
+This assumes all React code lives under a directory called `client`, the entry
````
Isn't the convention to put it in src (or src/client)? Also I don't think most people use .jsx extensions.

tgandrews replied:
I don't mind on this one. @cody do you have a preference?

brigand commented on docs/docs/getting-started.md:
````diff
@@ -41,9 +41,63 @@ $ npm install --save react react-dom babel-preset-react
+
+To run the above use the following command
+```sh
+webpack -p --config webpack.config.js
````
webpack.config.js is the default --config.

brigand commented on docs/docs/getting-started.md:
```diff
@@ -41,9 +41,63 @@ $ npm install --save react react-dom babel-preset-react
+For development mode things you may wish to do include:
+ - Removing the Uglify plugin
+ - Including source maps by adding the
+ [devtool](https://webpack.github.io/docs/configuration.html#devtool) option.
```
Maybe mention that 'eval' is a good devtool

andreypopp replied:
Actually `cheap-module-eval-source-map` as `eval` doesn't work with transforms which don't preserve line numbers (Babel). Also `cheap-module-eval-source-map` should be the default in Webpack 2 according to [docs](https://webpack.github.io/docs/configuration.html#devtool).

brigand replied:
Interesting, well time to update my webpack configs, thanks for the tip :-)

andreypopp commented on docs/docs/getting-started.md:
````diff
@@ -41,9 +41,63 @@ $ npm install --save react react-dom babel-preset-react
+webpack -p --config webpack.config.js
+```
+
+For development mode things you may wish to do include:
````
Maybe having a single `webpack.config.js` is better as it allows to get started with development quickly:
* For dev we can ask to run webpack with `--devtool` option
* For prod we can ask to run webpack `-p` option (which activates Uglify plugin).

tgandrews replied:
Is it worth including this in what is just a starting point and I did not expect it to be a final config. Hence, it just comments what could be done for dev mode and does not include a comprehensive config.

andreypopp replied:
My understanding that after reading docs people could start developing apps with React and Webpack immediately. Docs shouldn't be comprehensive but minimal, provide essential and sane dev and prod setup. Devtool is really an essential things as we use Babel/JSX.

andreypopp replied:
What I mean is "here's the config", for dev run "webpack --devtool ...", when ready to build for prod run "webpack -p ...".

andreypopp commented on docs/docs/getting-started.md:
```diff
@@ -41,9 +41,63 @@ $ npm install --save react react-dom babel-preset-react
+    filename: 'bundle.js',
+  },
+  resolve: {
+    modulesDirectories: ['node_modules', 'client'],
```
Having `client` in `modulesDirectories` seems like arbitrary convention to me.

andreypopp replied:
... which complicates configuration by the way. The same for `.jsx` actually but I'm less opinionated about this.

tgandrews replied:
I don't mind on this one, I use `jsx` for React components and then `js` for all pure JS files. The `client` convention as well is something I have used in a few projects. I am happy to change this.

gaearon replied:
I agree with @andreypopp, it seems more magic than it actually is. I'd just leave `['node_modules']` there so it's clear how to add more folders though.

zpao replied:
Agreed with @andreypopp & @gaearon

joshhunt replied:
I know I'm late to this, but `['node_modules']` is the default anyway, so that whole `resolve: {}` block could be removed (presuming `.jsx` is dropped in favour for `.js`), making the config less overwhelming for a 'getting started, and the webpack docs covers this config and many others in depth.

tgandrews commented:
I will squash the commits once the discussion is complete.

gaearon commented on docs/docs/getting-started.md:
```diff
@@ -37,13 +37,68 @@ $ browserify -t [ babelify --presets [ react ] ] main.js -o bundle.js
+  module: {
+    loaders: [{
+      test: /\.jsx?$/,
+      exclude: /node_modules/,
```
Rather than using `exclude` it's better to use `include: path.join(__dirname__, 'src')` or equivalent because this doesn't break for npm link'd modules.

zpao commented on docs/docs/getting-started.md:
````diff
@@ -37,13 +37,68 @@ $ browserify -t [ babelify --presets [ react ] ] main.js -o bundle.js
+```javascript
+module.exports = {
+  entry: [
+    './client/index.jsx',
````
Let's just make this `main.js` since that what we called the code above and put in the browserify example

zpao commented on docs/docs/getting-started.md:
````diff
@@ -37,13 +37,68 @@ $ browserify -t [ babelify --presets [ react ] ] main.js -o bundle.js
->
-> If you are using ES2015, you will want to also use the `babel-preset-es2015` package.
+A good starting webpack config for production mode would be:
+```javascript
````
We'll need to make sure we have whitespace surrounding code fences otherwise this gets dropped inline (I encourage you to run our site locally to ensure it looks right - directions are in the readme in the docs folder).

zpao commented:
My original thinking here was that we would do something more like #5686 for this page and then have more in depth documentation on a separate page. With this as is (and the markdown formatting fixed), it's a really daunting page with a lot of scary information that is not necessary to "get started". Here's the full rendered page: https://cloudup.com/cOdkbVBRUoY

So I'm inclined to just take the other PR and then we can flesh out a full page on using bundlers (or 2 separate pages). Does that sound reasonable to people?

gaearon commented:
We also shouldn't show the config for production mode separately from the development config. People will just copy whatever's on the page, and end up with a build that has no source maps, no propTypes checking, etc.

tgandrews commented:
:+1: I agree. This needs a separate page where you can go more in depth.

ryneeverett commented:
> This needs a separate page where you can go more in depth.

Putting a production config in the "getting started" page and a development config in a separate page strikes me as backwards.

gaearon commented:
Thank you very much for the effort.

Ultimately it seems that the result is indeed too in-depth for a getting started page. We merged #5686 and as a starting point, and there is some work in progress on separate bundler docs in #6148. I’ll mention your PR there so it stays for the reference in case we want to cherry-pick some parts from it.

Cheers!
